### PR TITLE
fix(node): Node 10 support

### LINKED ISSUE #5477: Replace use of deprecated/unsafe new Buffer constructor
**What is the current behavior?**

`yarn` uses deprecated `new Buffer()` constructor and causes deprecation warnings when run with `NODE_PENDING_DEPRECATION=1`.

```
$ ag '\bBuffer\('
src/registries/npm-registry.js
340:        const pw = new Buffer(String(password), 'base64').toString();
341:        return 'Basic ' + new Buffer(String(username) + ':' + pw).toString('base64');

src/util/fs.js
835:const cr = new Buffer('\r', 'utf8')[0];
836:const lf = new Buffer('\n', 'utf8')[0];
```

**What is the expected behavior?**

`yarn` should not use deprecated/unsafe `Buffer` constructor. According to the deprecation warning `new Buffer()` should be replaced with one of `Buffer.alloc()`, `Buffer.allocUnsafe()` or `Buffer.from()`; the [`safe-buffer`](https://yarn.pm/safe-buffer) package is another option.

**Please mention your node.js, yarn and operating system version.**

```
$ node -v
v8.9.4

$ yarn -v
1.5.1

$ uname -a
Linux 4.15.6-300.fc27.x86_64 #1 SMP Mon Feb 26 18:43:03 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

### LINKED ISSUE #5761: Unable to install local tgz file after upgraded to node 10
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**
Install local tgz file either hang at 'Resolving packages' or exit after 'Fetching packages'

**If the current behavior is a bug, please provide the steps to reproduce.**

I've created this [test package](https://www.dropbox.com/s/yn7zcqbewk5edte/yarn-tgz.tar.gz?dl=1). Just run `yarn add file:../b.tgz` in package `a`. b.tgz is packed by `yarn pack`.

**What is the expected behavior?**
tgz file gets unpacked and installed.

**Please mention your node.js, yarn and operating system version.**
yarn: 1.6.0
nodejs: 10.0.0
OS: macOS 10.12.6

**More details.**
The problem started right after I upgrade to node 10, before it worked fine.

Sometimes yarn hang at 'Resolving packages'
```
yarn add v1.6.0
(node:1191) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
verbose 0.544 Checking for configuration file "/private/tmp/yarn/a/.npmrc".
verbose 0.544 Checking for configuration file "/Users/quangan/.npmrc".
verbose 0.545 Found configuration file "/Users/quangan/.npmrc".
verbose 0.545 Checking for configuration file "/usr/local/etc/npmrc".
verbose 0.546 Checking for configuration file "/private/tmp/yarn/a/.npmrc".
verbose 0.546 Checking for configuration file "/private/tmp/yarn/.npmrc".
verbose 0.546 Checking for configuration file "/private/tmp/.npmrc".
verbose 0.546 Checking for configuration file "/private/.npmrc".
verbose 0.549 Checking for configuration file "/private/tmp/yarn/a/.yarnrc".
verbose 0.549 Checking for configuration file "/Users/quangan/.yarnrc".
verbose 0.549 Found configuration file "/Users/quangan/.yarnrc".
verbose 0.549 Checking for configuration file "/usr/local/etc/yarnrc".
verbose 0.55 Checking for configuration file "/private/tmp/yarn/a/.yarnrc".
verbose 0.55 Checking for configuration file "/private/tmp/yarn/.yarnrc".
verbose 0.55 Checking for configuration file "/private/tmp/.yarnrc".
verbose 0.55 Checking for configuration file "/private/.yarnrc".
verbose 0.552 current time: 2018-04-30T16:59:20.229Z
info No lockfile found.
[1/4] 🔍  Resolving packages...
⢀ file:../b.tgz
```

Sometimes it dies after 'Fetching packages'
```
yarn add v1.6.0
(node:5003) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
verbose 0.261 Checking for configuration file "/tmp/app/.npmrc".
verbose 0.261 Checking for configuration file "/Users/quangan/.npmrc".
verbose 0.261 Checking for configuration file "/usr/local/etc/npmrc".
verbose 0.261 Checking for configuration file "/tmp/app/.npmrc".
verbose 0.262 Checking for configuration file "/tmp/.npmrc".
verbose 0.263 Checking for configuration file "/tmp/app/.yarnrc".
verbose 0.263 Checking for configuration file "/Users/quangan/.yarnrc".
verbose 0.263 Found configuration file "/Users/quangan/.yarnrc".
verbose 0.263 Checking for configuration file "/usr/local/etc/yarnrc".
verbose 0.263 Checking for configuration file "/tmp/app/.yarnrc".
verbose 0.263 Checking for configuration file "/tmp/.yarnrc".
verbose 0.264 Found configuration file "/Users/quangan/.yarnrc".
verbose 0.264 Checking for configuration file "/Users/.yarnrc".
verbose 0.265 current time: 2018-04-27T05:11:53.672Z
[1/4] 🔍  Resolving packages...
verbose 0.401 Performing "GET" request to "https://registry.yarnpkg.com/uuid".
verbose 0.545 Request "https://registry.yarnpkg.com/uuid" finished with status code 200.
[2/4] 🚚  Fetching packages...
[#################################################################################################################################################################################################################################] 834/835
```

There is no yarn-error.log file.

### PR DESCRIPTION
**Pull request by imsnif**
**Summary**

1. Added node 10 to CircleCI
2. Fixed a test that was using the `net.createServer` method. This was having issues, likely because of this: https://github.com/nodejs/node/issues/20334 - it now uses a "normal" `http` server instead, serving the same purpose.
3. Changed the `readFirstAvailableStream` method in `util/fs.js`. I'm guessing this is related to the same issue as 2, but there have been quite some changes in `10` that might be causing this - so it's just a guess. Since this piece of code could be clearer and more conceise anyway, I opted to just change it. Now instead of relying on the stream `error` event to check if a file exists, it just explicitly checks it before starting the stream.
4. Upgraded `upath` from `1.0.2` to `1.0.5` which supports node 10 (found through its `engines` field when installing).
5. Upgraded Jest
6. Fixed a bug regarding replacing symlinks to non-existent targets on Windows
7. Removed a redundant test
8. Fixed two test cases that were failing frequently on macOS builds

fixes #5761
fixes #5477

**Test plan**

CI should be green (including and especially the new node10 job!)

### CODE REVIEW

**Comment by arcanis** [comment]
We should remove `ref` and anything that require it from the tests - it's a native dependency, so it's no surprise it breaks for some versions of Node :/

**Comment by leosco** [comment]
@imsnif have you updated your usage of `Buffer()` to `Buffer.from()` instead? Since upgrading to Node v10, every `yarn` command I run causes this warning to be printed:

```
(node:14277) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
```

Unless there's a problematic edge case I'm not aware of, you could probably do a quick find/replace and include that commit in this PR :smile:

**Comment by BYK** [comment]
@leosco this was already done.

**Comment by matt3224** [comment]
When will this move to a release?

**Comment by BYK** [comment]
@matt3224 once we figure out the issue with Windows builds, we'll merge this PR and release a new version.

**Comment by kachkaev** [comment]
@BYK any news? 🙏

**Comment by arcanis** [comment]
`replace the symlink when it changes, when using the link: protocol` is still failing on Windows. If someone on this platform can give us a hand, it would be super helpful :)

**Comment by theandrewlane** [comment]
@BYK @aracarie - Can we clear the AppVeyor build cache and rerun this build? 🤞🙏

I don't have a Windows machine, otherwise I would help!

**Comment by BYK** [comment]
Good news everyone! I've traced the issue down to a change in how `fs.realpath` works on Windows with symlinks pointing to non-existent targets. I'm working on a fix and will also likely to report this to the Node community if it wasn't reported already.

File: __tests__/commands/install/integration.js [hunk 1]
```diff
@@ -886,74 +782,61 @@ test.concurrent('should skip integrity check and do install when --skip-integrit
     // force rewrites lockfile
     newLockContent = await fs.readFile(path.join(config.cwd, 'yarn.lock'));
     expect(lockContent).not.toEqual(newLockContent);
-  });
-});
-
-test.concurrent('should install if symlink source does not exist', async (): Promise<void> => {
```
[the remark this answers is not in the document]
Comment (rally25rs) [reply __tests__/commands/install/integration.js]: oh nevermind, I see this was part of a commit "remove redundant test"

**Comment by rally25rs** [comment]
fyi, I added a couple "fixes" references to the original PR text so we can get a couple node10 specific issues closed out when this merges.

File: __tests__/fixtures/install/install-file-without-cache/comp/.gitignore [hunk 2]
```diff
@@ -1,2 +0,0 @@
-# this file will be generated
-index.js
```
[the remark this answers is not in the document]
Comment (BYK) [reply __tests__/fixtures/install/install-file-without-cache/comp/.gitignore]: Yes. This file was dynamically written right before two test runs which I believe was responsible for the test failures on macOS. See https://github.com/yarnpkg/yarn/pull/5769/commits/53923929e536cbcb1e71db8e09e2cafcac3addd2 for the full change related to this.

**Pull request merged**